### PR TITLE
chore: update node engine and .nvmrc version

### PR DESCRIPTION
**Pull request by wgwz**
## Description

When I was running things locally I was having trouble.
Specifying the node version in `.nvmrc` as more modern version is helpful I think.
Without updating the `package.json` engine value, I was seeing an error when I tried running `yarn`.
Heroku uses that engine value too though, so need to make sure that heroku still works with that change.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] targeted `dev` branch
- [ ] provided a link to the relevant issue or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed
- [ ] submit a PR against `master` branch after this one (and other PRs depending on this one, e.g. on regen-network/regen-web, if applicable) get merged

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please **add
your handle next to the items reviewed if you only reviewed selected items**.*

I have...

- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed code correctness and readability
- [ ] reviewed documentation is accurate
- [ ] reviewed tests
- [ ] manually tested (if applicable)

### CODE REVIEW

**Review by paul121 — COMMENTED** [review]
Thanks @wgwz was just making this change locally too, the latest graphile-worker v0.16 drops support for Node < 18: https://worker.graphile.org/releases#v0160

We should also update the docker image to Node 18

**Comment by paul121** [comment]
Hmm although I'm getting this issue when trying to POST to the metadata graph w/ Node 18.

> jsonld.InvalidUrl: Dereferencing a URL did not result in a valid JSON-LD object. Possible causes are an inaccessible URL perhaps due to a same-origin policy (ensure the server uses CORS if you are using client-side JavaScript), too many redirects, a non-JSON response, or more than one HTTP Link Header was provided for a remote context.

```shell
curl -H "Content-Type: application/json" --data @test.json http://localhost:5000/data/v1/iri-gen
```

```json
{
  "@context": "http://schema.org/",
  "description": "The Empire State Building is a 102-story landmark in New York City.",
  "geo": {
    "latitude": {
      "type": "xsd:float",
      "@value": "40.75"
    },
    "longitude": {
      "type": "xsd:float",
      "@value": "73.98"
    }
  },
  "image": "http://www.civil.usherbrooke.ca/cours/gci215a/empire-state-building.jpg",
  "name": "The Empire State Building"
}
```

**Comment by paul121** [comment]
Ah, fixed by upgrading the [`jsonld`](https://github.com/digitalbazaar/jsonld.js) dependency

**Comment by paul121** [comment]
@wgwz I added a couple commits. Any idea why tests aren't running? Is it because this is a PR from your fork?

**Comment by wgwz** [comment]
> Any idea why tests aren't running? Is it because this is a PR from your fork?

Ah, yeah I suppose that's the problem.. I think it may be that the github action is only configured for running on push to branches within this repo:

https://github.com/regen-network/regen-server/blob/dev/.github/workflows/registryServer.yml#L5

At some point in time I think we had PRs configured here, and I may have actually been the one to remove those.. thought they were unnecessary duplication. But maybe they'd be helpful to add back in... they'd probably need to have a filter that excludes all the dependabot stuff, because that duplication would be annoying (imagine running all the tests for both pushes and pull requests to those branches  😵‍💫 )

Here's how to exclude branches:
https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-excluding-branches

I think this is relevant too, if we turn on the workflow for PRs again, it might require approval from maintainers to run those:
https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks

But maybe this is all silly anyways.. perhaps i could have just pushed a branch to the parent repo 😆 I wasn't sure if that would work, which is why I used a fork in the first place

**Comment by paul121** [comment]
Yeah interesting @wgwz ... I think it would be nice to enable tests on PR for open source contributions. In my work with FarmOS we've just let there be duplicate action runs in this scenario and not worried about it but we don't have as much noise from dependabot.

For now I've just pushed and opened #454 including these changes and got tests to run.

But there is a failure in the iri-gen tests, seeming related to a new ["Safe mode"](https://github.com/digitalbazaar/jsonld.js?tab=readme-ov-file#safe-mode) in jsonld: https://github.com/regen-network/regen-server/actions/runs/8411971999/job/23032217210

Safe mode seems like a good thing to keep turned on because it is extra validation against "lossy behavior" (good since we are making content hashes) but I wonder if this might be a breaking change. Is it possible that JSON documents that are being POSTed to the iri-gen endpoint (project data etc from regen-web) might be "lossy" and would start triggering this error? It would be good to catch these but also don't want to break the frontend application cc @blushi

**Comment by blushi** [comment]
> I think it would be nice to enable tests on PR for open source contributions.

I believe this should do the trick: https://github.com/regen-network/regen-server/pull/455

**Comment by blushi** [comment]
> Yeah interesting @wgwz ... I think it would be nice to enable tests on PR for open source contributions. In my work with FarmOS we've just let there be duplicate action runs in this scenario and not worried about it but we don't have as much noise from dependabot.
> 
> For now I've just pushed and opened #454 including these changes and got tests to run.
> 
> But there is a failure in the iri-gen tests, seeming related to a new ["Safe mode"](https://github.com/digitalbazaar/jsonld.js?tab=readme-ov-file#safe-mode) in jsonld: https://github.com/regen-network/regen-server/actions/runs/8411971999/job/23032217210
> 
> Safe mode seems like a good thing to keep turned on because it is extra validation against "lossy behavior" (good since we are making content hashes) but I wonder if this might be a breaking change. Is it possible that JSON documents that are being POSTed to the iri-gen endpoint (project data etc from regen-web) might be "lossy" and would start triggering this error? It would be good to catch these but also don't want to break the frontend application cc @blushi

Right now, the failing test is just because the error message is not the right one, this is when testing the iri gen with some invalid JSON-LD, so we could just update the error message to match the new one I guess.
I've tested the frontend project creation flow and jsonld pkg upgrade doesn't seem to have any side effect.

**Comment by paul121** [comment]
Hmm after fixing the `iri-gen` tests now the `server` tests are failing:

```
server: Waiting for localhost:5000.................................................................................................................................................................................................................................................
server: Timeout
server: error Command failed with exit code 1.
```

I split up the `yarn build` and `yarn serve` commands/steps so that their logs would be printed. I'm seeing an error related to missing `options.clientID`: https://github.com/regen-network/regen-server/actions/runs/8529575254/job/23365630380#step:10:53

File: .github/workflows/registryServer.yml [hunk 1]
```diff
@@ -51,17 +51,17 @@ jobs:
           POSTGRES_USER: postgres
           POSTGRES_PASSWORD: password
         options: >-
-          --health-cmd pg_isready
+          --health-cmd "pg_isready -d server -U postgres"
```
Comment (paul121) [inline .github/workflows/registryServer.yml:54]: This resolves some error messages that are happening in the postgres container from health checks. Not a biggie but easy fix I came across & implemented while debugging others

example: https://github.com/regen-network/regen-server/actions/runs/8519190219/job/23332717100#step:20:14

**Comment by paul121** [comment]
> I'm seeing an error related to missing options.clientID

Ah, then I got another error related to S3 credentials... This is all happening because this PR is not from a branch in this repository and the github action secrets are not available. I've pushed these same changes into a new PR and the tests are passing #461 

>  I split up the yarn build and yarn serve commands/steps so that their logs would be printed. 

I think this change is still useful because it at least shows the errors for why the application might not be starting

**Comment by blushi** [comment]
@paul121 Should we close this in favor of https://github.com/regen-network/regen-server/pull/461#pullrequestreview-1975148495 then?

Just a side note, currently the file upload tests rely on the AWS creds to pass, I believe a better option would be to mock any calls to AWS.

**Comment by paul121** [comment]
> @paul121 Should we close this in favor of #461 then?

:+1: 
 
> Just a side note, currently the file upload tests rely on the AWS creds to pass, I believe a better option would be to mock any calls to AWS.

Yes that makes sense

**Pull request closed without merging**